util/util.cpp

```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
+                            sensorTempFile >> sensorRealTemp;
+                            sensorTempFile.close();
+                        }
                     }
-                    else
+                    else if (sensorList[i][t->first].type == "file")
                     {
-                        sensorTempFile.close();
-                        continue;
+                        std::fstream sensorValueFile;
+                        sensorValueFile.open(sensorList[i][t->first].path, std::ios::in);
+                        if (sensorValueFile)
+                        {
+                            sensorValueFile >> sensorRealTemp;
+                            sensorValueFile.close();
+                        }
+                    }
+
+                    if (!(std::isnan(sensorRealTemp)))
+                    {
+                        // If configured unit not already degrees, convert to degrees
+                        if (incomingMilli)
+                        {
+                            sensorRealTemp /= 1000.0;
+                        }
                     }
-                    
-                    sensorTempFile.close();
-                }
-                else if (sensorList[i][t->first].type == "dbus")
-                {
-                    sensorRealTemp = 
-                        getSensorDbusTemp(sensorList[i][t->first].path);
                 }
-                else if (sensorList[i][t->first].type == "file")
+
+                if (std::isnan(sensorRealTemp))
                 {
-                    std::fstream sensorValueFile;
-                    sensorValueFile.open(sensorList[i][t->first].path, std::ios::in);
-                    if (sensorValueFile)
+                    // Sensor failure, unable to get reading
+                    if constexpr (DEBUG)
                     {
-                        sensorValueFile >> sensorRealTemp;
+                        std::cerr << " ?";
                     }
 
-                    sensorValueFile.close();
+                    continue;
                 }
 
-                if (sensorList[i][t->first].unit == "degree")
+                // If sensor already in margin, then accept it as-is
+                if (incomingMargin)
                 {
-                    sensorRealTemp *= 1000;
-                    sensorSpecTemp *= 1000;
-                }
-                else if (sensorList[i][t->first].unit == "millimargin"){
-                    if (calibMarginTemp == -1 ||
+                    // Remember this margin if it is the worst margin
+                    if ((std::isnan(calibMarginTemp)) ||
                         sensorRealTemp < calibMarginTemp)
                     {
                         calibMarginTemp = sensorRealTemp;
                     }
+
+                    // Sensor successful, already in margin format
+                    if constexpr (DEBUG)
+                    {
+                        std::cerr << " " << sensorRealTemp;
```

Comment:
Error message place holder?

Reply:
This, and all the similar comments, are part of building up a line of output. If you run the full loop until you get to the end, it will build up output like this: "Margin Zone 0: 90 86 90 85 58.875 78.875 58.5 60.75 55.312 57.437 57.062 60.312 55.437 56.562 => 55.312"
In this we can see all of the incoming margins from the temperature sensors, and the worst margin decision that was made. If an individual sensor malfunctions, it will show up as "?" instead of a number, so the total number of items in the output will always remain the same.

Reply:
I see. Can we construct this logging with a string (either sstream or std::string::operator+=, but I'm not sure which one has less performance overhead), so that this log will not be interrupted by some log message in some function call in the middle, and it's more readable?

Reply:
Yes, this sounds entirely reasonable. Makes sense, as in case somebody wants to add some debugging, it will interrupt the current line being built up. Should build up an internal string then log the entire line at once. Will add it as a future refinement.